code/obj/item/device/audio_log.dm

```diff
@@ -290,88 +229,183 @@
+				speak_message = copytext(speak_message, separator+1)
+
+				speak(speaker, speak_message)
+				sleep(5 SECONDS)
+				if (!tape || !tape.next(continuous))
+					stop()
+
+	proc/stop()
+		src.mode = MODE_OFF
+		src.updateSelfDialog()
+		if (src.self_destruct)
+			SPAWN(2 SECONDS)
+				src.explode()
+
+	proc/speak(speaker, message, show_no_speaker, text_colour_override)
+		if (!message)
 			return
+		var/speaker_colour
+		if (!text_colour_override)
+			speaker_colour = src.text_colour
+			if (speaker && !show_no_speaker && name_colours[speaker])
+				speaker_colour = name_colours[speaker]
+		else
+			speaker_colour = text_colour_override
+		if (!speaker && !show_no_speaker)
+			speaker = "Unknown"
+
+		var/image/chat_maptext/audio_log_text
+		if (istype(src.loc, /turf))
+			audio_log_text = make_chat_maptext(src, message, "color: [speaker_colour];")
+			if (audio_log_text && src.chat_text && length(src.chat_text.lines))
+				audio_log_text.measure(src)
+				for (var/image/chat_maptext/I in src.chat_text.lines)
+					if (I != audio_log_text)
+						I.bump_up(audio_log_text.measured_height)
+		src.audible_message("<span class='game radio' style='color: [speaker_colour]'><span class='name'>[speaker]</span><b> [bicon(src)]\[Log\]</b> <span class='message'>\"[message]\"</span></span>", 2, assoc_maptext = audio_log_text)
+		return
 
-		var/speaker = copytext(speak_message, 1, separator)
-		speak_message = copytext(speak_message, separator+1)
+	proc/explode()
+		speak(null, "This message will self-destruct in 5 seconds...", TRUE, "#E00000")
+		sleep(1 SECOND)
+		for (var/i in 1 to 4)
+			speak(null, "[5 - i]", TRUE, "#E00000")
+			sleep(1 SECOND)
 
-		src.speak(speaker, speak_message)
-		if (!src.tape.next(continuous))
-			src.mode = 0
-			processing_items.Remove(src)
-			src.updateSelfDialog()
+		src.blowthefuckup(2)
 		return
 
+	proc/create_name_colours(var/list/names)
+		if (!names.len)
+			return
 
-	proc
-		speak(speaker, message)
-			if(!message)
-				return
-			if(!speaker)
-				speaker = "Unknown"
+		var/list/unique_names = list()
+		for (var/i in 1 to names.len)
+			if (!(names[i] in unique_names))
+				unique_names.Add(names[i])
 
-			for(var/mob/O in all_hearers(5, src.loc))
-				O.show_message("<span class='game radio'><span class='name'>[speaker]</span><b> [bicon(src)]\[Log\]</b> <span class='message'>\"[message]\"</span></span>",2)
+		name_colours = list()
+		if (unique_names.len == 1)
+			name_colours[unique_names[1]] = text_colour
 			return
 
-		explode()
+		var/list/text_rgb = hex_to_rgb_list(text_colour)
+		var/list/text_hsl = rgb2hsl(text_rgb[1], text_rgb[2], text_rgb[3])
+		var/lightness_part = 60 / (unique_names.len + 1)
 
-			var/turf/T = get_turf(src.loc)
+		for (var/i in 1 to unique_names.len)
+			var/lightness = 20 + (lightness_part * i)
+			var/colour = hsl2rgb(text_hsl[1], text_hsl[2], lightness)
+			name_colours[unique_names[i]] = colour
+			message_admins("[lightness]")
```

Comment:
🤨

Reply:
I did make a note to myself reminding me to remove all the debug code...